File: models/Gpt-3.5-turbo/cpp/code/problems/tree_algorithms/Fixed-Length_Paths_I.cpp

#include <iostream>
#include <vector>

using namespace std;

vector<vector<int>> adj; // adjacency list representation of the tree
vector<int> visited; // visited array to keep track of visited nodes
int pathCount = 0; // counter variable to count the paths

void dfs(int node, int pathLength) {
    visited[node] = true;
    
    if (pathLength == k) {
        pathCount++;
        return;
    }
    
    for (int neighbor : adj[node]) {
        if (!visited[neighbor]) {
            dfs(neighbor, pathLength + 1);
        }
    }
    
    visited[node] = false; // backtrack: mark the current node as unvisited
}

int main() {
    int n, k;
    cin >> n >> k;
    
    // Initialize the adjacency list and visited array
    adj.resize(n + 1);
    visited.resize(n + 1, false);
    
    // Read the edges of the tree
    for (int i = 0; i < n - 1; i++) {
        int a, b;
        cin >> a >> b;
        adj[a].push_back(b);
        adj[b].push_back(a);
    }
    
    // Start the DFS from any node
    dfs(1, 0);
    
    cout << pathCount << endl;
    
    return 0;
}

File: models/Gpt-3.5-turbo/py/code/tree_algorithms/Fixed-Length_Paths_I.py

def dfs(node, adj_list, visited, path_length, k, path_count):
    visited[node] = True

    # Check if the current path length matches k
    if path_length == k:
        path_count[0] += 1

    # Continue DFS on adjacent nodes
    for neighbor in adj_list[node]:
        if not visited[neighbor]:
            dfs(neighbor, adj_list, visited, path_length + 1, k, path_count)

    visited[node] = False

n, k = map(int, input().split())
adj_list = [[] for _ in range(n+1)]

# Build adjacency list
for _ in range(n-1):
    a, b = map(int, input().split())
    adj_list[a].append(b)
    adj_list[b].append(a)

# Initialize variables
visited = [False] * (n+1)
path_count = [0]

# Start DFS from each node
for node in range(1, n+1):
    dfs(node, adj_list, visited, 0, k, path_count)

print(path_count[0])
